ci/snapcraft.sh

```diff
@@ -0,0 +1,135 @@
+#!/bin/sh
+
+set -e
+
+# This script encapsulates the job of crafting snaps of rustup and publishing
+# them.  We construct either an edge snap (builds on PRs or master)
+# or a beta snap if built from a non-master branch.
+
+# If built on master, we will publish the edge snap
+# If built on stable, we will publish the beta snap
+
+# To parameterise the build, the following environment variables are needed:
+# DO_SNAP=1  <-- without this, we don't run
+# SNAP_EDGE=1 <-- with this, we build an edge/devel snap, otherwise a release snap
+# SNAP_ARCH=... <-- without this we cannot built, this is the snap architecture we're building
+
+if ! [ -n "$DO_SNAP" ]; then
+  echo "SNAP: Not attempting to snap"
+  exit 0
+fi
+
+if ! hash snapcraft >/dev/null 2>/dev/null; then
+  echo "SNAP: Unable to find snapcraft, cannot run"
+  exit 1
+fi
+
+if ! [ -n "$SNAP_ARCH" ]; then
+  echo "SNAP: Unable to generate snap, architecture unset"
+  exit 1
+else
+  echo "SNAP: Constructing snap for architecture $SNAP_ARCH"
+fi
+
+if [ -n "$SNAP_EDGE" ]; then
+  echo "SNAP: Generated snap will be an grade:devel snap"
+  GRADE=devel
+else
+  echo "SNAP: Generated snap will be a grade:stable snap"
+  GRADE=stable
+fi
+
+
+# This is the list of proxies which the snap will contain.  This needs to match
+# the set of aliases which the snap has been allocated.  If we add to this
+# list but don't get the aliases added, then people won't see the proxy except
+# as rustup.$PROXY.  If we have an alias which is not a supported proxy name
+# then rustup might get sad.
+
+PROXIES="cargo cargo-clippy cargo-fmt cargo-miri clippy-driver rls rustc rustdoc rustfmt"
+
+# From now on, things should be automagic
+
+VERSION=$(grep "^version" Cargo.toml | head -1 | cut -d\" -f2)
+
+rm -rf snap
+
+mkdir -p snap/snap
```

Comment:
I would prefer to rename this to `mkdir -p snapbuild/snap` to avoid repeatedly name.

Reply:
I think I picked `snap/snap` because of how snapcraft uses `snap` but of course it's not necessary at the top level, so I'll fix that.